cmd/dockerd/daemon.go

```diff
@@ -228,37 +231,32 @@ func (cli *DaemonCli) start() {
 		}
 		l, err := listeners.Init(proto, addr, serverConfig.SocketGroup, serverConfig.TLSConfig)
 		if err != nil {
-			logrus.Fatal(err)
+			return err
 		}
 		// If we're binding to a TCP port, make sure that a container doesn't try to use it.
 		if proto == "tcp" {
 			if err := allocateDaemonPort(addr); err != nil {
-				logrus.Fatal(err)
+				return err
 			}
 		}
 		logrus.Debugf("Listener created for HTTP on %s (%s)", protoAddrParts[0], protoAddrParts[1])
 		api.Accept(protoAddrParts[1], l...)
 	}
 
 	if err := migrateKey(); err != nil {
-		logrus.Fatal(err)
+		return err
 	}
 	cli.TrustKeyPath = cli.commonFlags.TrustKey
 
 	registryService := registry.NewService(cli.Config.ServiceOptions)
 	containerdRemote, err := libcontainerd.New(cli.getLibcontainerdRoot(), cli.getPlatformRemoteOptions()...)
 	if err != nil {
-		logrus.Fatal(err)
+		return err
 	}
 
 	d, err := daemon.NewDaemon(cli.Config, registryService, containerdRemote)
 	if err != nil {
-		if pfile != nil {
```

Comment:
Oh right, I was concerned about the `defer` being wrapped in the body of the `if` but it's indeed correct, thanks :+1: